### ISSUE TITLE
Swiftlint: closure body length

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/firefox-ios#16442]
Add `closue_body_length` to `swiftlint.yaml` and fix warnings in project resulting from this by moving code into separate functions.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-7301)

**Comment by ionixjunior** [comment]
I'd like to work on it.

**Comment by adudenamedruby** [comment]
@ionixjunior amazing. Assigned. :)

**Comment by ionixjunior** [comment]
Thank you, @adudenamedruby! I'm looking forward to it!

**Comment by ionixjunior** [comment]
Hello! Quick update about this issue: I've been working on adding the `closure_body_length` config and addressing the new violations. Initially found 49 violations, and I've resolved some of them - currently down to 29 remaining. I've been keeping my branch updated with the main branch, so everything is in sync.

Thanks for your patience.

**Comment by ionixjunior** [comment]
@adudenamedruby I've created the pull request. Please let me know if you need anything or if there are any changes I should make.

**Comment by ionixjunior** [comment]
After feedback from @adudenamedruby, we decided to close my big pull request, and split this work in smaller pull requests. My initial plan is to create one pull request per file changed. In the table below, I've listed all the files, and I'll update the status column to track the progress of this task. 

| File | Status |
|---|---|
| BrowserKit/Sources/SiteImageView/ImageProcessing/SiteImageFetcher/SiteImageDownloader.swift | ✅ |
| Account/FxAPushMessageHandler.swift | ✅ |
| Client/Application/UITestAppDelegate.swift | ✅ |
| Client/ContentBlocker/ContentBlocker.swift | ✅ |
| Client/Experiments/Experiments.swift | ✅ |
| Client/Frontend/Autofill/Address/AddressAutofillToggle.swift | ✅ |
| Client/Frontend/Autofill/Address/AddressListView.swift | ✅ |
| Client/Frontend/Autofill/Address/Edit/EditAddressViewController.swift | ✅ |
| Client/Frontend/Autofill/CreditCard/CreditCardSettingsView/CreditCardInputView.swift | ✅ |
| Client/Frontend/Autofill/CreditCard/CreditCardSettingsView/CreditCardItemRow.swift | ✅ |
| Client/Frontend/Autofill/CreditCard/CreditCardSettingsView/CreditCardSettingsEmptyView.swift | ✅ |
| Client/Frontend/Browser/Authenticator.swift | ✅ |
| Client/Frontend/Browser/BrowserViewController/Actions/ActionProviderBuilder.swift | ✅ |
| Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift | ✅ |
| Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift | ✅ |
| Client/Frontend/Browser/SearchLoader.swift | ✅ |
| Client/Frontend/Browser/SearchSuggestClient.swift | ✅ |
| Client/Frontend/Browser/Tabs/Legacy/LegacyTabPeekPreviewActionBuilder.swift | ✅ |
| Client/Frontend/Browser/Tabs/Legacy/LegacyTabPeekViewController.swift | ✅ |
| ~Client/Frontend/Browser/Toolbars/Redux/AddressBarState.swift~ |  |
| Client/Frontend/Browser/Toolbars/Redux/NavigationBarState.swift | ✅ |
| Client/Frontend/Browser/Toolbars/Redux/ToolbarState.swift | ✅ |
| Client/Frontend/Fakespot/FakespotState.swift | ✅ |
| Client/Frontend/Library/HistoryPanel/HistoryPanel.swift | ✅ |
| Client/Frontend/Library/Reader/ReaderPanel.swift | ✅ |
| ~Client/Frontend/Library/Reader/ReaderPanelEmptyStateView.swift~ |  |
| Client/Frontend/Microsurvey/Prompt/MicrosurveyPromptState.swift | ✅ |
| Client/Frontend/Reader/ReaderModeHandlers.swift | ✅ |
| Client/Frontend/Settings/ContentBlockerSettingViewController.swift | ✅ |
| Client/Frontend/Settings/ThemeSettings/ThemeSettingsState.swift | ✅ |
| Client/Frontend/TabContentsScripts/UserScriptManager.swift | ✅ |
| Providers/RustSyncManager.swift | ✅ |
| ~Storage/Rust/RustAutofill.swift~ |  |
| Storage/Rust/RustLogins.swift | ✅ |
| WidgetKit/ImageButtonWithLabel.swift | ✅ |
| WidgetKit/OpenTabs/OpenTabsWidget.swift | ✅ |
| WidgetKit/TopSites/TopSitesWidget.swift | ✅ |
| firefox-ios-tests/Tests/StorageTests/RustAutofillTests.swift | ✅ |
| ~firefox-ios-tests/Tests/XCUITests/FxScreenGraph.swift~ |  |
| .swiftlint.yml | ✅ |

**Comment by ionixjunior** [comment]
@adudenamedruby I sent the first PR focusing on improvements to the first file. It might be a bit large, but this is an exception. The other cases will be easier to review. Let me know if I need to change anything or split it into smaller chunks.

**Comment by data-sync-user** [comment]
➤ Matt Reagan commented:

Internal refactor, no QA needed

**Comment by data-sync-user** [comment]
➤ Roux Buciu commented:

This remains in progress, Nishant Bhasin  . There’s still instances of this in the code, and it cannot yet be closed as it's not yet in swiftlint.yaml. ionixjunior continues the awesome work of bringing us closer to getting this done.

**Comment by ionixjunior** [comment]
@FilippoZazzeroni I think I found an issue in [this PR](https://github.com/mozilla-mobile/firefox-ios/pull/22424). While it introduces improvements in build times by modifying the linter to analyze only changed files, I've noticed a problem during testing. The current command `MODIFIED_FILES=$(git diff --name-only --diff-filter=ACM | grep -e '\.swift$')` only captures unstaged changes. Once files are committed, the command returns nothing.

This could compromise the feedback from the linter in Xcode, as some new warnings might slip through unnoticed. Does that make sense to you? 

I suggest customizing the script for two scenarios:
1. For the feature branches, check differences from the main branch.
2. For the main branch, check differences in the most recent commit.

Here's a suggestion for the script:

```sh
if [[ "$(uname -m)" == arm64 ]]; then
    export PATH="/opt/homebrew/bin:$PATH"
fi

CURRENT_BRANCH=$(git branch --show-current)

if [[ "$CURRENT_BRANCH" == "main" ]]; then
    MODIFIED_FILES=$(git show --name-only --diff-filter=ACM | grep -e '\.swift$')
else
    MODIFIED_FILES=$(git diff --name-only --diff-filter=ACM main $CURRENT_BRANCH | grep -e '\.swift$')
fi

if which swiftlint > /dev/null; then
    for file in $MODIFIED_FILES; do
        swiftlint lint "${SRCROOT}/../${file}" --config "${SRCROOT}/../.swiftlint.yml"
    done
else
    echo "warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint" 
fi
```

In this script, I get the current branch name and, if it is the main branch, I use the `git show` command to get files changed on the most recent commit. Since the merge commit includes all changed files in a PR, this should cover everything. If we're not on the main branch, I use `git diff` against the main branch to capture the relevant files.

This might not be the perfect solution, but I believe it will improve the development experience using linter when working on feature branches. Let me know if this makes sense or if I missed anything.

**Comment by FilippoZazzeroni** [comment]
Hi @ionixjunior thanks you for the proposed upgrade, unfortunately i had to revert other build script that contained the same logic, for this one i will keep it like it is for now since it's going to need revision as well.
The change you are proposing makes sense, but it skips unstaged files from the feature branch. At the moment i will keep it as is since there is also from recently a githook who runs linter before committing and fail the commit when there are lints left.

**Comment by ionixjunior** [comment]
You're right, @FilippoZazzeroni. I overlooked the unstaged files. Thank you for your clarification, and I appreciate your feedback! If this becomes an issue in the future, feel free to reach out, because I enjoy solving "Git challenges".

**Comment by FilippoZazzeroni** [comment]
You're welcome @ionixjunior. Sure things i'll reach out when there is something that comes out, also if there are any new issue regarding that field.

**Comment by ionixjunior** [comment]
Nice, thank you!

**Comment by FilippoZazzeroni** [comment]
You're welcome @ionixjunior 😃

**Comment by ionixjunior** [comment]
@adudenamedruby @FilippoZazzeroni following [this PR](https://github.com/mozilla-mobile/firefox-ios/pull/23710), the focus will shift to addressing the most significant violations and gradually lowering the warning and error thresholds until all violations are resolved. The table created [here](https://github.com/mozilla-mobile/firefox-ios/issues/16442#issuecomment-2197676804) remains relevant, but I'm prioritizing the largest cases first (some of which are new and not mapped in the table).

**Comment by FilippoZazzeroni** [comment]
Hi @ionixjunior make total sense to me. Yes if we can address larger class first then we are able to lower faster the number.
Thanks again for your great work 🎉

**Comment by ionixjunior** [comment]
@FilippoZazzeroni the next step involves refactoring the `AddressBarState`. While we started this effort in PR https://github.com/mozilla-mobile/firefox-ios/pull/22943, it wasn't sufficient to reduce the closure body length to a standard value. Currently, the `reducer` property spans 53 lines, which is above the recommended threshold.

Upon analyzing the code, I noticed that the switch statement handles a variety of cases, but they can be grouped into three main responsibilities:

- Responding to toolbar changes.
- Responding to web page changes.
- Responding to search changes.

I came to this conclusion by testing the app, observing the actions triggered, and correlating them with the behaviors in the app. To address this, I propose splitting the switch statement into these three logical groups. This approach would reduce the overall size of the closure and help us achieve our objective.

Here’s a possible implementation for the final code:

```swift
static let reducer: Reducer<Self> = { state, action in
    guard action.windowUUID == .unavailable || action.windowUUID == state.windowUUID
    else {
        return defaultState(from: state)
    }

    if let result = handleToolbarActions(state: state, action: action) {
        return result
    }

    if let result = handleWebPageActions(state: state, action: action) {
        return result
    }

    if let result = handleSearchActions(state: state, action: action) {
        return result
    }

    return defaultState(from: state)
}

private static func handleToolbarActions(state: Self, action: Action) -> Self? {
    switch action.actionType {
    case ToolbarActionType.didLoadToolbars:
        return handleDidLoadToolbarsAction(state: state, action: action)

    case ToolbarActionType.numberOfTabsChanged:
        return handleNumberOfTabsChangedAction(state: state, action: action)

    case ToolbarActionType.traitCollectionDidChange:
        return handleTraitCollectionDidChangeAction(state: state, action: action)

    case ToolbarActionType.showMenuWarningBadge:
        return handleShowMenuWarningBadgeAction(state: state, action: action)

    case ToolbarActionType.toolbarPositionChanged:
        return handlePositionChangedAction(state: state, action: action)

    case ToolbarActionType.borderPositionChanged:
        return handlePositionChangedAction(state: state, action: action)

    case ToolbarActionType.hideKeyboard:
        return handleHideKeyboardAction(state: state)

    case ToolbarActionType.cancelEditOnHomepage:
        return handleCancelEditOnHomepageAction(state: state, action: action)

    default:
        return nil
    }
}

private static func handleWebPageActions(state: Self, action: Action) -> Self? {
    switch action.actionType {
    case ToolbarActionType.readerModeStateChanged:
        return handleReaderModeStateChangedAction(state: state, action: action)

    case ToolbarActionType.websiteLoadingStateDidChange:
        return handleWebsiteLoadingStateDidChangeAction(state: state, action: action)

    case ToolbarActionType.urlDidChange:
        return handleUrlDidChangeAction(state: state, action: action)

    case ToolbarActionType.backForwardButtonStateChanged:
        return handleBackForwardButtonStateChangedAction(state: state, action: action)

    default:
        return nil
    }
}

private static func handleSearchActions(state: Self, action: Action) -> Self? {
    switch action.actionType {
    case ToolbarActionType.didPasteSearchTerm:
        return handleDidPasteSearchTermAction(state: state, action: action)

    case ToolbarActionType.didStartEditingUrl:
        return handleDidStartEditingUrlAction(state: state, action: action)

    case ToolbarActionType.cancelEdit:
        return handleCancelEditAction(state: state, action: action)

    case ToolbarActionType.didSetTextInLocationView:
        return handleDidSetTextInLocationViewAction(state: state, action: action)

    case ToolbarActionType.clearSearch:
        return handleClearSearchAction(state: state, action: action)

    case ToolbarActionType.didDeleteSearchTerm:
        return handleDidDeleteSearchTermAction(state: state, action: action)

    case ToolbarActionType.didEnterSearchTerm:
        return handleDidEnterSearchTermAction(state: state, action: action)

    case ToolbarActionType.didSetSearchTerm:
        return handleDidSetSearchTermAction(state: state, action: action)

    case ToolbarActionType.didStartTyping:
        return handleDidStartTypingAction(state: state, action: action)

    case SearchEngineSelectionActionType.didTapSearchEngine:
        return handleDidTapSearchEngine(state: state, action: action)

    case SearchEngineSelectionMiddlewareActionType.didClearAlternativeSearchEngine:
        return handleDidClearAlternativeSearchEngine(state: state, action: action)

    default:
        return nil
    }
}
```

I acknowledge that this change could be somewhat controversial and might introduce mistakes, but I believe it’s a worthwhile direction to explore.

## Pros
- Maintains the closure body length within acceptable limits.
- Groups code logically by actions or changes, improving contextual clarity.

## Cons
- It could be challenging to classify new states when adding additional functionality.
- Might obscure the full set of available options, depending on how the code is structured.
- Even with significant effort, adding new states to the "search" or "toolbar" responsibilities may still lead to challenges, as these sections are already quite extensive.

I’m open to any feedback, suggestions, or alternative approaches to help tackle this issue effectively.

**Comment by FilippoZazzeroni** [comment]
Thanks so much @ionixjunior for this extensive report i like this idea and i'd go for it, only one slight change, i'd redirect to the sub method only if the action is of the correct type otherwise i'd skip it.

Something like the following
```swift
static let reducer: Reducer<Self> = { state, action in
        // Only process actions for the current window
        guard action.windowUUID == .unavailable || action.windowUUID == state.windowUUID
        else {
            return defaultState(from: state)
        }

        if let action  = action as? TabTrayAction {
            return reduceTabTrayAction(action: action, state: state)
        } else if let action = action as? TabPanelMiddlewareAction {
            return reduceTabPanelMiddlewareAction(action: action, state: state)
        } else if let action = action as? TabPanelViewAction {
            return reduceTabPanelViewAction(action: action, state: state)
        }

        return defaultState(from: state)
    }
```

**Comment by ionixjunior** [comment]
@FilippoZazzeroni thank you for the suggestion! I reviewed the action types in the `reducer`property, and I noticed that most of them belong to the `ToolbarAction` type. Using this approach, the property still spans about 46 lines.

I liked your "type-oriented" approach, as it promotes better organization. However, since we primarily deal with just two types - `ToolbarAction` and `SearchEngineSelectionAction` - this introduces a bit of a challenge.

Based on your suggestion, I drafted the following idea:

```swift
static let reducer: Reducer<Self> = { state, action in
    guard action.windowUUID == .unavailable || action.windowUUID == state.windowUUID
    else {
        return defaultState(from: state)
    }

    if let action = action as? ToolbarAction {
        return reduceToolbarAction(state: state, action: action)
    } else if let action = action as? SearchEngineSelectionAction {
        return reduceSearchEngineSelectionAction(state: state, action: action)
    }

    return defaultState(from: state)
}

private static func reduceToolbarAction(state: Self, action: Action) -> Self {
    guard let actionType = action.actionType as? ToolbarActionType else { return defaultState(from: state) }

    switch actionType {
    case .didLoadToolbars: return handleDidLoadToolbarsAction(state: state, action: action)
    case .numberOfTabsChanged: return handleNumberOfTabsChangedAction(state: state, action: action)
    case .traitCollectionDidChange: return handleTraitCollectionDidChangeAction(state: state, action: action)
    case .showMenuWarningBadge: return handleShowMenuWarningBadgeAction(state: state, action: action)
    case .toolbarPositionChanged: return handlePositionChangedAction(state: state, action: action)
    case .borderPositionChanged: return handlePositionChangedAction(state: state, action: action)
    case .hideKeyboard: return handleHideKeyboardAction(state: state)
    case .cancelEditOnHomepage: return handleCancelEditOnHomepageAction(state: state, action: action)
    case .readerModeStateChanged: return handleReaderModeStateChangedAction(state: state, action: action)
    case .websiteLoadingStateDidChange: return handleWebsiteLoadingStateDidChangeAction(state: state, action: action)
    case .urlDidChange: return handleUrlDidChangeAction(state: state, action: action)
    case .backForwardButtonStateChanged: return handleBackForwardButtonStateChangedAction(state: state, action: action)
    case .didPasteSearchTerm: return handleDidPasteSearchTermAction(state: state, action: action)
    case .didStartEditingUrl: return handleDidStartEditingUrlAction(state: state, action: action)
    case .cancelEdit: return handleCancelEditAction(state: state, action: action)
    case .didSetTextInLocationView: return handleDidSetTextInLocationViewAction(state: state, action: action)
    case .clearSearch: return handleClearSearchAction(state: state, action: action)
    case .didDeleteSearchTerm: return handleDidDeleteSearchTermAction(state: state, action: action)
    case .didEnterSearchTerm: return handleDidEnterSearchTermAction(state: state, action: action)
    case .didSetSearchTerm: return handleDidSetSearchTermAction(state: state, action: action)
    case .didStartTyping: return handleDidStartTypingAction(state: state, action: action)
    default: return defaultState(from: state)
    }
}

private static func reduceSearchEngineSelectionAction(state: Self, action: Action) -> Self {
    switch action.actionType {
    case SearchEngineSelectionActionType.didTapSearchEngine:
        return handleDidTapSearchEngine(state: state, action: action)

    case SearchEngineSelectionMiddlewareActionType.didClearAlternativeSearchEngine:
        return handleDidClearAlternativeSearchEngine(state: state, action: action)

    default:
        return defaultState(from: state)
    }
}
```

In this implementation, I focused on `ToolbarAction`and `SearchEngineSelectionAction`. For toolbar-related actions, I ensured the type of `actionType` is `ToolbarActionType` to keep the `case` statements concise, with return statements on the same line. Unfortunately, I couldn't apply the same pattern to the search engine reducer because it involves multiple action types - `SearchEngineSelectionActionType` and `SearchEngineSelectionMiddlewareActionType`. 

With these changes, the toolbar reducer will span 26 lines.

Let me know your thoughts and if you'd suggest any further adjustments.

**Comment by FilippoZazzeroni** [comment]
Thanks so much @ionixjunior i understand this approach i fill only it is a bit too much condensed. I think either if we can do sub groups or also i think we can call 48 lines as a good results for the closure body 👀 , what do you think >

**Comment by ionixjunior** [comment]
@FilippoZazzeroni you're absolutely right, the last solution I proposed does look overly condensed, and I didn’t like it much either when revisiting it. Regarding the closure body length, I’m a bit hesitant about defining 48 lines as acceptable for this rule. While I agree that exceptions can and should exist, I believe they shouldn't become the default behavior. Based on the current refactor I'm doing, it seems there might only be three or four cases like this across the project, so they could all be treated as exceptions rather than a standard.

That said, I prefer the idea of sub-groups to keep things organized. However, I also think it’s important to gather more opinions on this to better evaluate the trade-offs and gain different perspectives from the team. Perhaps we could involve others in this discussion to ensure a well-rounded approach?

One way to define these sub-groups is to revisit [the first proposal](https://github.com/mozilla-mobile/firefox-ios/issues/16442#issuecomment-2605060094). The downside, as you mentioned, is the additional sub-function calls for each group until the correct condition is found.

Alternatively, we could temporarily do nothing about the length for now in this file. While this might sound weird, it could allow us to focus on continuing to reduce the threshold in standard cases. We could disable the closure body length rule for this specific property using inline comments and proceed with the refactor until the rule value aligns more closely with the standard value. Once the refactor is complete, we’d have a clearer understanding of these exceptions and could make a more informed final decision based on the overall state of the codebase.

What do you think? I'm happy to adjust to the team's direction.

**Comment by FilippoZazzeroni** [comment]
Hi @ionixjunior i had a look better to the initial proposal and the reason why a was leaning to the type oriented structure was because i thought we could lose information by returning earlier from the reducer method, anyway looking better this can't happen. I like the idea of sub groups, and i would not treat those closure as exception, since disabling rule for those could lead to increase their size. Anyway tagging @adudenamedruby @OrlaM to get more feedback on this [approach](https://github.com/mozilla-mobile/firefox-ios/issues/16442#issuecomment-2605060094), i believe we could go for it since it does logically encapsulate actions, we should discuss tho the:
```swift
if let result = handleAction {
   return result
}
```
Does it align with the current way we implement reducers ?

Thanks so much @ionixjunior, again great work 🎉

**Comment by ionixjunior** [comment]
Hello @FilippoZazzeroni. No problem, and thanks for following up! That sounds like a solid plan. The main goal now is to further reduce the overall size of this property, so feel free to assign those new tickets to me. Once they're solved, I'll continue decreasing this Swiftlint rule.

**Comment by FilippoZazzeroni** [comment]
Thanks so much @ionixjunior, yes i'm going to create the tickets today.

**Comment by ionixjunior** [comment]
After discussing the best approach for refactoring `AddressBarState`, we decided to temporarily disable the `closure_body_length`rule for this closure using comment. This allows us to proceed with other refactors and continue reducing the overall threshold across the project.

We plan to revisit this closure later to remove the comment and implement a final solution. However, this will likely require a more robust approach that demands additional planning and time.

Thank you so much, @FilippoZazzeroni, for your support and insightful discussions on Element!

**Comment by ionixjunior** [comment]
Hello everyone. This week, I’ll continue working on this refactoring. Apologies for the delay, but rest assured, we’ll complete it.

**Comment by FilippoZazzeroni** [comment]
Hi @ionixjunior no problem, there is no rush 😃

**Comment by ionixjunior** [comment]
@adudenamedruby @FilippoZazzeroni after merging the last pull request, this issue was closed. We've reduced the closure body length to 34 lines. The biggest that I solved was 212 lines. We made a good improvement comparing to the beginning.

@mattreaganmozilla raised a good point [here](https://github.com/mozilla-mobile/firefox-ios/pull/27605#discussion_r2172699371). It seems we’ve already achieved the objective. Do you think we should continue working toward the standard closure body length of 30 lines, or are we good with the current progress? 

To achieve 30 lines we need to solve more 19 violations.

**Comment by FilippoZazzeroni** [comment]
Hi @ionixjunior i personally think that 34 lines it's really a great result and we should be good with it. I think the other area we could focus would be the `function_body_length` with the same approach for the closure body, as we still don't support it.

**Comment by ionixjunior** [comment]
Sure! Thank you so much for your guidance and patience throughout this journey @FilippoZazzeroni!

**Comment by FilippoZazzeroni** [comment]
No problem @ionixjunior thank you for all the great work
